File: 2025-12/2025-12-04/permutationCount.js
function countPermutations(str) {
  // P = n! / (n1! * n2! * ... * nk!)
  const letterSet = new Set(str);
  const numerator = factorial(str.length);
  let denominator = 1;
  
  for (const letter of letterSet) {
    const letterCount = str.match(new RegExp(letter, "g")).length;
    denominator *= factorial(letterCount);
  }

  return numerator / denominator;
}

const factorial = n => n <= 1 ? 1 : n * factorial(n - 1);

// --- TEST SUITE ---

const testsText = `
1. countPermutations("abb") should return 3.
2. countPermutations("abc") should return 6.
3. countPermutations("racecar") should return 630.
4. countPermutations("freecodecamp") should return 39916800.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+)\.$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a)
    if (a[index] !== b[index])
      return false;
  
  return true;
}

runTests(testData);

File: 12 December/12-04/permutation_count.py
def count_permutations(s):
    # P = n! / (n1! * n2! * ... * nk!)
    letter_set = set(s)
    numerator = factorial(len(s))
    denominator = 1

    for letter in letter_set:
        letter_count = len(re.findall(letter, s))
        denominator *= factorial(letter_count)

    return numerator / denominator

def factorial(n):
    return 1 if n <= 1 else n * factorial(n - 1)

# --- TEST SUITE ---

tests_text = '''
1. count_permutations("abb") should return 3.
2. count_permutations("abc") should return 6.
3. count_permutations("racecar") should return 630.
4. count_permutations("freecodecamp") should return 39916800.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+)\.$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)